### PR TITLE
Refactor handling of MIR bodies

### PR DESCRIPTION
**Pull request by voidc**
This enables translating const fns.

### CODE REVIEW

File: creusot/src/callbacks.rs [hunk 1]
```diff
@@ -23,6 +43,27 @@ impl Callbacks for ToWhy {
...
+                // We use `mir_promoted` as it is the MIR required by borrowck
+                let (body, promoted) = tcx.mir_promoted(ty::WithOptConstParam::unknown(def_id));
+                let body_and_proms = BodyAndPromoteds {
+                    body: body.borrow().clone(),
+                    promoted: promoted.borrow().clone(),
+                };
+
+                // SAFETY: The reader casts the 'static lifetime to 'tcx before using it.
+                let body_and_proms: BodyAndPromoteds<'static> =
+                    unsafe { std::mem::transmute(body_and_proms) };
```
Comment (xldenis) [inline creusot/src/callbacks.rs:57]: uh. is there no better way to do this? (I can't think of any off the top of my head)

File: creusot/src/ctx.rs [hunk 2]
```diff
@@ -22,26 +23,44 @@ use rustc_errors::{DiagnosticBuilder, DiagnosticId};
...
 use why3::exp::Exp;
 
 pub(crate) use crate::translated_item::*;
 
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub struct BodyId {
+    def_id: LocalDefId,
+    pub promoted: Option<Promoted>,
+}
+
```
Comment (xldenis) [inline creusot/src/ctx.rs:46]: FINALLY, GOOD, THANKS. Incidentally, I've seen discussions about giving promoteds their own `DefId`s but this is good.

File: creusot/src/callbacks.rs [hunk 3]
```diff
@@ -53,3 +94,16 @@ impl Callbacks for ToWhy {
         }
     }
 }
+
+pub fn get_body<'tcx>(tcx: TyCtxt<'tcx>, def_id: LocalDefId) -> Option<BodyAndPromoteds<'tcx>> {
```
Comment (xldenis) [inline creusot/src/callbacks.rs:98]: I think this is worth a small function level comment, explaining that it fetches from a thread local catch and will panic and will return `None` if called a second time.

File: creusot/src/translation/function/promoted.rs [hunk 4]
```diff
@@ -24,13 +23,13 @@ pub(crate) fn promoted_signature<'tcx>(body: &Body<'tcx>) -> PreSignature<'tcx>
 
 pub(crate) fn translate_promoted<'tcx>(
     ctx: &mut TranslationCtx<'tcx>,
-    body: &Body<'tcx>,
-    parent: DefId,
+    body_id: BodyId,
 ) -> CreusotResult<(PreSignature<'tcx>, fmir::Body<'tcx>)> {
-    let func_translator = BodyTranslator::build_context(ctx.tcx, ctx, &body, parent);
+    let body = ctx.body(body_id).clone();
```
Comment (xldenis) [inline creusot/src/translation/function/promoted.rs:28]: I wish we didn't have to clone the body here but if that turns out to be a perf issue later, we can always either wrap it in an `Rc` or use interior mutability in `ctx` like rustc does.

**Comment by xldenis** [comment]
lgtm

File: creusot/src/callbacks.rs (hunk 1 above)
Comment (voidc) [reply creusot/src/callbacks.rs:57]: it's kind of stupid but I don't see any other way since storage in the thread local requires `'static`

**Pull request merged**